Add rwx test cases

khushboo-rancher opened the pull request:
https://github.com/longhorn/longhorn/issues/1606

joshimoo commented on manager/integration/tests/test_rwx.py:
```diff
@@ -1,12 +1,358 @@
+
+
+@pytest.fixture(scope="module", autouse="True")
+def nfs_pvc(request):
```
The creation and cleanup of these resources is really nice.
Great job 👍

joshimoo commented on manager/integration/tests/test_rwx.py:
```diff
@@ -29,7 +375,46 @@ def test_rwx_with_statefulset_multi_pods():
+
+    create_and_wait_statefulset(statefulset)
+
+    command = "ls /export | grep -i 'pvc' | wc -l"
```
nice approach for checking this data :)

joshimoo commented on manager/integration/tests/test_rwx.py:
```diff
@@ -1,12 +1,358 @@
+            tty=False)
+
+
+@pytest.mark.skip(reason="Waiting for rwx enhancement")  # NOQA
```
Do we still need to skip the test?

khushboo-rancher replied:
Created the PR https://github.com/longhorn/longhorn-manager/pull/671, I think once that is merged we can remove the skip.

joshimoo reviewed (commented):
LGTM, do we still need to skip the test?

khushboo-rancher commented:
Added the resources needed in test environment to execute rwx tests and removed the skip.
Added one more test ```test_rwx_multi_statefulset_with_same_pvc```

khushboo-rancher commented:
Added tests ```test_rwx_parallel_writing, test_rwx_statefulset_scale_down_up, test_rwx_deployment_with_multi_pods```

joshimoo commented on manager/integration/tests/test_rwx.py:
```diff
@@ -1,12 +1,373 @@
+            ],
+            'resources': {
+                'requests': {
+                    'storage': '20G'
```
For the test environment you can use a tiny storage amount, for the longhorn volume.

joshimoo commented on manager/integration/tests/test_rwx.py:
```diff
@@ -1,12 +1,373 @@
+    wait_deployment_replica_ready(apps_api, 'longhorn-nfs-provisioner', 1,
+                                  namespace='longhorn-system')
+    # This is to make sure nfs server is started.
+    time.sleep(25)
```
There is a probably a better way of checking for this, but this is okay for now.

joshimoo commented on manager/integration/tests/test_rwx.py:
```diff
@@ -89,11 +544,63 @@ def test_rwx_parallel_writing():
+
+    with Pool(2) as p:
+        p.map(write_data_into_pod, [pod_name, statefuleset_pod_name])
+
```
I think you need to wait for completion of the above parallel tasks?

khushboo-rancher replied:
Only after the parallel write is completed, it goes to the next line to execute.

joshimoo commented on manager/integration/tests/test_rwx.py:
```diff
@@ -119,11 +626,82 @@ def test_rwx_statefulset_scale_down_up():
 @pytest.mark.skip(reason="TODO")
-def test_rwx_offline_node_longhorn_nfs_provisioner():
+def test_rwx_offline_node_longhorn_nfs_provisioner(
+    core_api, nfs_storage_class, statefulset, pod):  # NOQA
```
all the other functions params are indented an additional step

khushboo-rancher replied:
Fixed it.

joshimoo reviewed (dismissed):
LGTM in general

meldafrawi commented on manager/integration/tests/test_rwx.py:
```diff
@@ -1,12 +1,358 @@
+
+
+@pytest.fixture(scope="module", autouse="True")
+def core_api(request):
```
We already have `core_api` fixture defined in `common.py`

meldafrawi replied:
After discussing with @khushboo-rancher
`core_api` and `apps_api` defined in `common.py` is function scoped fixtures
It is pytest limitation that function scoped fixture can't be used in higher scoped fixtures

One minor change we can introduce, is to rename module scoped `core_api`  to something like `core_api_module` so it can be easily differentiated from function scoped `core_api` in `common.py`.  it's totally not to change it though.

khushboo-rancher replied:
Changed the name to ```core_api_module```

meldafrawi commented:
LGTM

khushboo-rancher commented:
@joshimoo @meldafrawi Made the changes as suggested. Please review.

yasker commented:
Can you merge the commits into one? Otherwise LGTM.

khushboo-rancher commented:
@yasker Merged the commits into one.
